Add GitHub actions runner to ci-tasks container

martinpitt opened the pull request:
This allows us to start a self-hosted runner pretty much anywhere,
including locally (for debugging) or on internal infrastructure (for
RHEL tests).

Include a `github-action-run-once` script that does everything that's
required for this, so that the container invocation becomes fairly
simple (see comment at the top of the file).

 - [x] Builds on top of PR #2899, which changes the entrypoint; land that first

martinpitt commented:
I tested this with an [extra commit](https://github.com/martinpitt/anaconda/commit/a58b98735e9ce00f06207f7bdacf2f6c333b0d77) that essentially changes `runs-on: ubuntu-latest` to `runs-on: ["self-hosted", "X64", "ci-tasks"]` (plus some extra noise). I started a local runner with

    podman run -h pittirunner1 -e GITHUB_TOKEN=$(< ~/.config/github-token) -e GITHUB_REPOSITORY=martinpitt/anaconda  --rm anaconda/ci-tasks github-action-run-once

which then appeared on [my fork's actions](https://github.com/martinpitt/anaconda/settings/actions):

![image](https://user-images.githubusercontent.com/200109/95190334-7192c180-07cf-11eb-8de6-1428d17e1fe7.png)

The local container picks this up, runs that one request, the output [gets published on my repo's Actions](https://github.com/martinpitt/anaconda/runs/1214254712?check_suite_focus=true); then the container exits cleanly, and the runner gets deregistered:

```
√ Connected to GitHub

2020-10-06 10:26:31Z: Listening for Jobs
2020-10-06 10:26:41Z: Running job: unit-tests
2020-10-06 10:35:38Z: Job unit-tests completed with result: Succeeded

# Runner removal

√ Runner removed successfully
√ Removed .credentials
√ Removed .runner
```

In a real deployment, systemd would now start a fresh container for the next request.

This provides the necessary building blocks for my proposed schema how to run PRs on a self-hosted runner in a reasonably isolated manner, except that the systemd unit does not exist yet (I just started the container by hand for the first test):

![sequence](https://www.websequencediagrams.com/files/render?link=hauYoSBKikUsP24J3YCxvGRL6LuhZbn2pErEO8B0Vam23vvwPDOULY28JndI3kiw)

jkonecny12 commented on dockerfile/ci-tasks/github-action-run-once:
```diff
@@ -0,0 +1,37 @@
+fi
+
+# register runner
+./config.sh --labels 'ci-tasks' ${RUNNER_NAME:+--name} ${RUNNER_NAME:-} --replace --token "$RUNNER_TOKEN" --unattended --url "https://github.com/$GITHUB_REPOSITORY"
```
This is a bit crazy bashism to me but I was able to understand what is happening there.

martinpitt replied:
It's not a bashism, it's POSIX shell. That said, if you think it's weird, I'm happy to replace with an if condition and building some `$EXTRA_ARGS` or so. (That's why we have reviews, thanks for pointing out!)

VladimirSlavik replied:
This one seems self-explanatory to me, and I have not been exposed to shells too much... Then again maybe trusting that hunch is not such a great idea. It's shell after all...

jkonecny12 reviewed (approved):
Looks good to me. Thanks!

martinpitt commented:
**UPDATES**:
 - Rebase on  top of #2899, which changes from CMD to ENTRYPOINT. This changes the invocation of the container.
 - I updated the README and the detailled "how to test" comment above.
 - I also moved the default entrypoint from /root to / for convenience and consistency.

@jkonecny12 : thus nothing substantial, but do you want to re-review? (Also, should we configure github to drop approvals when the PR changes?)

martinpitt commented:
Rebased after landing #2899, no code changes.
